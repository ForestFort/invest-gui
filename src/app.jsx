import fs from 'fs';
import path from 'path';
import React from 'react';

import {InvestJob} from './InvestJob';
import { getInvestList } from './getInvestList'

const CACHE_DIR = 'cache' //  for storing state snapshot files

export default class App extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      investList: {},
      recentSessions: [],
      investSettings: {},
    };
    this.updateRecentSessions = this.updateRecentSessions.bind(this);
    this.saveSettings = this.saveSettings.bind(this);
  }

  async componentDidMount() {
    // TODO: also load and set investSettings from a cached state
    const investList = await getInvestList();
    const recentSessions = await findRecentSessions(CACHE_DIR);
    this.setState(
      {
        investList: investList,
        recentSessions: recentSessions,
        investSettings: {
          nWorkers: '-1',
          loggingLevel: 'INFO',
        }
      });
  }

  updateRecentSessions(sessionID) {
    // This triggers on saveState clicks
    let recentSessions = Object.assign([], this.state.recentSessions);
    recentSessions.unshift(sessionID);
    this.setState({recentSessions: recentSessions});
  }

  saveSettings(settings) {
    this.setState({
      investSettings: settings
    });
  }

  render() {
    return (
      <InvestJob 
        investList={this.state.investList}
        investSettings={this.state.investSettings}
        recentSessions={this.state.recentSessions}
        updateRecentSessions={this.updateRecentSessions}
        saveSettings={this.saveSettings}
      />
    );
  }
}

function findRecentSessions(cache_dir) {
  // Populate recentSessions from list of files in cache dir
  // sorted by modified time.

  // TODO: check that files are actually state config files
  // before putting them on the array
  return new Promise(function(resolve, reject) {
    const files = fs.readdirSync(cache_dir);

    // reverse sort (b - a) based on last-modified time
    const sortedFiles = files.sort(function(a, b) {
      return fs.statSync(path.join(cache_dir, b)).mtimeMs -
           fs.statSync(path.join(cache_dir, a)).mtimeMs
    });
    // trim off extension, since that is how sessions
    // were named orginally
    resolve(sortedFiles
      .map(f => path.parse(f).name)
      .slice(0, 15) // max 15 items returned
    );
  });
}

