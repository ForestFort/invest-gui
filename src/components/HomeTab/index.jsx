import React from 'react';
import PropTypes from 'prop-types';
import Electron from 'electron';
import path from 'path';

import Button from 'react-bootstrap/Button';
import ButtonGroup from 'react-bootstrap/ButtonGroup';
import CardGroup from 'react-bootstrap/CardGroup';
import Card from 'react-bootstrap/Card';
import Spinner from 'react-bootstrap/Spinner';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';

const STATUS_COLOR_MAP = {
  running: 'warning',
  error: 'danger',
  success: 'success'
}

export class HomeTab extends React.Component {
  /** Renders a button for each invest model and for each cached invest job.
  */

  constructor(props) {
    super(props);
  }

  render () {
    // A button for each model
    const investJSON = this.props.investList;
    let investButtons = [];
    for (const model in investJSON) {
      investButtons.push(
        <Button key={model}
          value={investJSON[model]['internal_name']}
          onClick={this.props.investGetSpec}
          variant="outline-success">
          {model}
        </Button>
      );
    }

    return (
      <Row>
        <Col md={6}>
          <ButtonGroup vertical className="mt-2">
            {investButtons}
          </ButtonGroup>
        </Col>
        <Col md={6}>
          <RecentInvestJobs
            loadState={this.props.loadState}
            recentSessions={this.props.recentSessions}/>
        </Col>
      </Row>
    );
  }
}

HomeTab.propTypes = {
  investList: PropTypes.object,
  investGetSpec: PropTypes.func,
  saveState: PropTypes.func,
  loadState: PropTypes.func,
  recentSessions: PropTypes.array
}


class RecentInvestJobs extends React.Component {
  /** Renders a button for each recent invest job. Also displays job metadata.
  *
  * Recent job metadata is passed here via props, originally loaded from
  * a persistent file when the app is launched.
  */
  
  constructor(props) {
    super(props);
    // this.selectFile = this.selectFile.bind(this);
    this.handleClick = this.handleClick.bind(this);
  }

  // selectFile(event) {
  //   const dialog = Electron.remote.dialog;
  //   // TODO: could add more filters to only show .json
  //   dialog.showOpenDialog({
  //     properties: ['openFile']
  //   }, (filepath) => {
  //     if (filepath[0]) {
  //       this.props.loadState(
  //         path.parse(path.basename(filepath[0])).name); // 0 is safe since we only allow 1 selection
  //     }
  //   })
  // }

  handleClick(sessionFilename) {
    this.props.loadState(sessionFilename);
  }

  render() {

    // Buttons to load each recently saved state
    let recentButtons = [];
    this.props.recentSessions.forEach(session => {
      const name = session[0];
      const model = session[1].model;
      const workspaceDir = session[1].workspace.directory;
      const suffix = session[1].workspace.suffix;
      const status = session[1]['status'];
      const description = session[1]['description'];
      const datetime = session[1]['humanTime'];

      recentButtons.push(
        <Card className="text-left session-card w-100"
          as="button"
          key={name}
          // value={session[1]['statefile']} 
          onClick={() => this.handleClick(session[1]['statefile'])}
          border={STATUS_COLOR_MAP[status] || 'dark'}>
          <Card.Body>
            <Card.Header>
              {model}  
              {status === 'running' && 
                <Spinner as='span' animation='border' size='sm' role='status' aria-hidden='true'/>
              }
            </Card.Header>
            <Card.Title>
              <em>Workspace: </em>{workspaceDir}
            </Card.Title>
            <Card.Title>
              { suffix && <em>Suffix: </em> }{suffix}
            </Card.Title>
            <Card.Text>{description || <em>no description</em>}</Card.Text>
          <Card.Footer className="text-muted">{datetime}</Card.Footer>
          </Card.Body>
        </Card>
      );
    });
    // Also a button to browse to a cached state file if it's not in recent list
    // recentButtons.push(
    //   <Button
    //     key="browse"
    //     type="submit"
    //     variant="secondary"
    //     onClick={this.selectFile}>
    //     Browse for saved session
    //   </Button>
    // );

    return (
      <React.Fragment>
        <div>
          Select Recent Session:
        </div>
        <CardGroup className='session-card-group'>
          {recentButtons}
        </CardGroup>
      </React.Fragment>
    );
  }
}

LoadStateForm.propTypes = {
  loadState: HomeTab.propTypes.loadState,
  recentSessions: HomeTab.propTypes.recentSessions
}