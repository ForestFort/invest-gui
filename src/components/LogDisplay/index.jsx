import React from 'react';

import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Alert from 'react-bootstrap/Alert';
import Button from 'react-bootstrap/Button';

const logStyle = {
  whiteSpace: 'pre-line',
  maxHeight: '700px',
  overflowY: 'scroll',
};

export class LogDisplay extends React.Component {

  constructor(props) {
    super(props);
    this.content = React.createRef();
  }

  componentDidUpdate() {
    this.content.current.scrollTop = this.content.current.scrollHeight;
  }

  render() {
    const current_err = this.props.logStdErr;
    // Include the stderr in the main log even though it also gets an Alert
    const current_out = this.props.logStdOut + current_err;
    let renderedLog;
    let renderedAlert;
    let killButton;

    renderedLog =
        <Col ref={this.content} style={logStyle}>
          {current_out}
        </Col>

    if (current_err) {
      renderedAlert = <Alert variant={'danger'}>{current_err}</Alert>
    } else {
      if (this.props.sessionProgress === 'viz') { // this was set if python exited w/o error
        renderedAlert = <Alert variant={'success'}>{'Model Completed'}</Alert>
      }
    }

    killButton = 
      <Button
        variant="primary" 
        size="lg"
        onClick={this.props.investKill}>
        Kill Subprocess
      </Button>

    return (
      <React.Fragment>
        <Row>{renderedLog}</Row>
        <Row>{renderedAlert}</Row>
      </React.Fragment>
    );
  }
}