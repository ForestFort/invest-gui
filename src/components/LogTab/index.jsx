import React from 'react';
import { Tail } from 'tail';
import path from 'path';
import os from 'os';

import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Container from 'react-bootstrap/Container';
import Alert from 'react-bootstrap/Alert';
import Button from 'react-bootstrap/Button';

const logStyle = {
  whiteSpace: 'pre-line',
  maxHeight: '700px',
  overflowY: 'scroll',
};

class LogDisplay extends React.Component {

  constructor(props) {
    super(props);
    this.content = React.createRef();
  }

  componentDidUpdate() {
    this.content.current.scrollTop = this.content.current.scrollHeight;
  }

  render() {
    return (
      <Col ref={this.content} style={logStyle}>
        {this.props.logdata}
      </Col>
      );
  }
}

export class LogTab extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      logdata: ''
    }
    this.tail = null;
  }

  componentDidUpdate(prevProps) {
    // if there is a logfile and it's new, start tailing the file.
    if (this.props.logfile && prevProps.logfile !== this.props.logfile) {
      this.tail = new Tail(this.props.logfile, {
        fromBeginning: true
      });
      let logdata = Object.assign('', this.state.logdata);
      this.tail.on('line', (data) => {
        logdata += `${data}` + os.EOL
        this.setState({ logdata: logdata })
      })

    // No new logfile. No existing logdata.
    } else if (this.state.logdata === '') {
      this.setState({logdata: 'Starting...'})

    // No new logfile. Existing logdata. Invest process exited. 
    } else if (['success', 'error'].includes(this.props.jobStatus)) {
      this.tail.unwatch()
    }
  }

  render() {    
    let renderedAlert;
    if (this.props.jobStatus === 'error') {
      renderedAlert = <Alert variant={'danger'}>{this.props.logStdErr}</Alert>
    } else if (this.props.jobStatus === 'success') {
      renderedAlert = <Alert variant={'success'}>{'Model Completed'}</Alert>
    }

    return (
      <Container>
        <Row>
          <LogDisplay logdata={this.state.logdata}/>
        </Row>
        <Row>{renderedAlert}</Row>
      </Container>
    );
  }
}