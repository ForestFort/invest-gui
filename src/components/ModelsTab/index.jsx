import path from 'path';
import React from 'react';
import Electron from 'electron';

import Button from 'react-bootstrap/Button';
import ButtonGroup from 'react-bootstrap/ButtonGroup';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';


export class ModelsTab extends React.Component {

  constructor(props) {
    super(props);
  }

  render () {
    // A button for each model
    const investJSON = this.props.investList;
    let investButtons = [];
    for (const model in investJSON) {
      investButtons.push(
        <Button key={model}
          value={investJSON[model]['internal_name']}
          onClick={this.props.investGetSpec}
          variant="outline-success">
          {model}
        </Button>
      );
    }

    return (
      <Row>
        <Col md={6}>
          <ButtonGroup vertical className="mt-2">
            {investButtons}
          </ButtonGroup>
        </Col>
        <Col md={6}>
          <Row className="mt-2">
            <LoadStateForm
              loadState={this.props.loadState}
              recentSessions={this.props.recentSessions}/>
          </Row>
        </Col>
      </Row>
    );
  }
}


class LoadStateForm extends React.Component {
  
  constructor(props) {
    super(props);
    this.state = {
      session_id_to_load: ''
    }
    this.selectFile = this.selectFile.bind(this);
  }

  selectFile(event) {
    const dialog = Electron.remote.dialog;
    // TODO: could add more filters based on argType (e.g. only show .csv)
    dialog.showOpenDialog({
      properties: ['openFile']
    }, (filepath) => {
      if (filepath[0]) {
        this.props.loadState(
          path.parse(path.basename(filepath[0])).name); // 0 is safe since we only allow 1 selection
      }
    })
  }

  render() {

    // Buttons to load each recently saved state
    let recentButtons = [];
    this.props.recentSessions.forEach(session => {
      recentButtons.push(
        <Button  className="text-left"
          key={session}
          value={session}
          onClick={this.handleLink}
          variant='outline-dark'>
          {session}
        </Button>
      );
    });
    // Also a button to browse to a cached state file if it's not in recent list
    recentButtons.push(
      <Button
        type="submit"
        variant="secondary"
        onClick={this.selectFile}>
        Browse for saved session
      </Button>
    );

    return (
      <div>
        <div>
          Select Recent Session:
        </div>
        <ButtonGroup vertical className="mt-2">
          {recentButtons}
        </ButtonGroup>
      </div>
    );
  }
}
