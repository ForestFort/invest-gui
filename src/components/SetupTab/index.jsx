import fs from 'fs';
import React from 'react';
import Electron from 'electron';
import request from 'request';

import Button from 'react-bootstrap/Button';
import Form from 'react-bootstrap/Form';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';
import InputGroup from 'react-bootstrap/InputGroup';

export class SetupTab extends React.Component {

  render () {

    // Only mount the ArgsForm when there are actually args
    // This lets us have an ArgsForm.componentDidMount() that
    // does useful initialization of args state.
    if (this.props.args) {
      return (
        <div>
          <ArgsForm 
            args={this.props.args}
            modulename={this.props.modulename}
            updateArg={this.props.updateArg}
            batchUpdateArgs={this.props.batchUpdateArgs}
            investValidate={this.props.investValidate}
            argsValuesFromSpec={this.props.argsValuesFromSpec}
          />
          <Button 
            variant="primary" 
            size="lg"
            onClick={this.props.investExecute}
            disabled={!this.props.argsValid}>
                Execute
          </Button>
        </div>);
    }
    // The SetupTab remains disabled in this route, so no need
    // to render anything here.
    return(<div>No args to see here</div>)
  }
}

class ArgsForm extends React.Component {

  constructor(props) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
    this.selectFile = this.selectFile.bind(this);
    this.onDragDrop = this.onDragDrop.bind(this);
  }

  componentDidMount() {
    // Validate args immediately on mount. Values will be `undefined`
    // but converted to empty strings by `argsValuesFromSpec` 
    // Validation will handle updating the `valid` property
    // of the optional and conditionally required args so that they 
    // can validate without any user-interaction.

    // TODO: could call batchUpdateArgs here instead
    // to avoid passing investValidate to this component at all.
    // this.props.batchUpdateArgs(JSON.parse(args_dict_string));
    this.props.investValidate(this.props.argsValuesFromSpec(this.props.args));
  }

  handleChange(event) {
    // Handle changes in form text inputs
    const value = event.target.value;
    const name = event.target.name;
    this.props.updateArg(name, value);
  }

  selectFile(event) {
    // Handle clicks on form browse-button inputs
    const dialog = Electron.remote.dialog;
    const argtype = event.target.value;
    const argname = event.target.name;
    const prop = (argtype === 'directory') ? 'openDirectory' : 'openFile'
    // TODO: could add more filters based on argType (e.g. only show .csv)
    dialog.showOpenDialog({
      properties: [prop]
    }, (filepath) => {
      this.props.updateArg(argname, filepath[0]); // 0 is safe since we only allow 1 selection
    })
  }

  onDragDrop(event) {
    // Handle drag-drop of datastack JSON files and InVEST logfiles
    event.preventDefault();
    
    const fileList = event.dataTransfer.files;
    if (fileList.length !== 1) {
      throw alert('only drop one file at a time.')
    }
    const filepath = fileList[0].path;
    request.post(
      'http://localhost:5000/post_datastack_file',
      { json: { 
        datastack_path: filepath} 
      },
      (error, response, body) => {
        if (!error) {
          const datastack = body;
          if (datastack['module_name'] === this.props.modulename) {
            this.props.batchUpdateArgs(datastack['args']);
          } else {
            throw alert('Parameter/Log file for ' + datastack['module_name'] + ' does not match this model: ' + this.props.modulename)
          }
        } else {
          console.log('Error: ' + error.message)
        }
      }
    );
  }

  render() {
    const current_args = Object.assign({}, this.props.args)
    let formItems = [];
    for (const argname in current_args) {
      if (argname === 'n_workers') { continue }
      const argument = current_args[argname];

      // These types need a text input and a file browser button
      if (['csv', 'vector', 'raster', 'directory'].includes(argument.type)) {
        formItems.push(
          <Form.Group as={Row} key={argname}>
            <Form.Label column sm="3">{argument.name}</Form.Label>
            <Col sm="8">
              <InputGroup>
                <Form.Control
                  name={argname}
                  type="text" 
                  value={argument.value || ''} // empty string is handled better than `undefined`
                  onChange={this.handleChange}
                  isValid={argument.touched && argument.valid}
                  isInvalid={argument.touched && argument.validationMessage}
                />
                <InputGroup.Append>
                  <Button 
                    variant="outline-secondary"
                    value={argument.type}  // dialog will limit options to files or dirs accordingly
                    name={argname}
                    onClick={this.selectFile}>
                    Browse
                  </Button>
                </InputGroup.Append>
                <Form.Control.Feedback type='invalid'>
                  {argument.type + ' : ' + (argument.validationMessage || '')}
                </Form.Control.Feedback>
              </InputGroup>
            </Col>
          </Form.Group>)
      
      // These types need a text input
      } else if (['freestyle_string', 'number'].includes(argument.type)) {
        formItems.push(
          <Form.Group as={Row} key={argname}>
            <Form.Label column sm="3">{argument.name}</Form.Label>
            <Col sm="4">
              <Form.Control
                name={argname}
                type="text" 
                value={argument.value || ''} // empty string is handled better than `undefined`
                onChange={this.handleChange}
                isValid={argument.touched && argument.valid}
                isInvalid={argument.touched && argument.validationMessage}
              />
              <Form.Control.Feedback type='invalid'>
                {argument.type + ' : ' + (argument.validationMessage || '')}
              </Form.Control.Feedback>
            </Col>
          </Form.Group>)
      
      // Radio select for boolean args
      } else if (argument.type === 'boolean') {
        // argument.value will be type boolean if it's coming from an args dict
        // generated by natcap.invest. But it will be type string if the value
        // is set from this UI, because html forms always submit strings.
        // So, `checked` property must accomodate both types to determine state.
        formItems.push(
          <Form.Group as={Row} key={argname}>
            <Form.Label column sm="3">{argument.name}</Form.Label>
            <Col sm="8">
              <Form.Check
                inline
                type="radio"
                label="Yes"
                value={"true"}
                checked={argument.value || argument.value === "true"}
                onChange={this.handleChange}
                name={argname}
              />
              <Form.Check
                inline
                type="radio"
                label="No"
                value={"false"}
                checked={!argument.value || argument.value === "false"}
                onChange={this.handleChange}
                name={argname}
              />
            </Col>
          </Form.Group>)

      // Dropdown menus for args with options
      } else if (argument.type === 'option_string') {
        formItems.push(
          <Form.Group as={Row} key={argname}>
            <Form.Label column sm="3">{argument.name}</Form.Label>
            <Col sm="4">
              <Form.Control
                as='select'
                name={argname}
                value={argument.value}
                onChange={this.handleChange}>
                {argument.validation_options.options.map(opt =>
                  <option value={opt} key={opt}>{opt}</option>
                )}
              </Form.Control>
              <Form.Control.Feedback type='invalid'>
                {argument.type + ' : ' + (argument.validationMessage || '')}
              </Form.Control.Feedback>
            </Col>
          </Form.Group>)
      }
    }

    return (
      <Form 
        validated={false}
        onDrop={this.onDragDrop}
        onDragOver={dragover_handler}>
        {formItems}
      </Form>
    );
  }
}

function dragover_handler(event) {
 event.preventDefault();
 event.dataTransfer.dropEffect = "move";
}