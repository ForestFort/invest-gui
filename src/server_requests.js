import request from 'request';

export function getSpec(payload) {
  return new Promise(function(resolve, reject) {
    request.post(
      'http://localhost:5000/getspec',
      payload,
      (error, response, body) => {
        if (!error && response.statusCode == 200) {
          resolve(body);
        } else {
          console.log('Status: ' + response.statusCode)
          console.log('Error: ' + error.message)
          return null;
        }
      }
    );
  });
}