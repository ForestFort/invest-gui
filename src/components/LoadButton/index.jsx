import React from 'react';
import Electron from 'electron'
import PropTypes from 'prop-types';

import Row from 'react-bootstrap/Row';
import Form from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';
import Modal from 'react-bootstrap/Modal';
import Dropdown from 'react-bootstrap/Dropdown';

import { fetchDatastackFromFile } from '../../server_requests';

export class LoadButton extends React.Component {
  /** Render a button that loads args from a datastack, parameterset, or logfile.
  * Opens an native OS filesystem dialog to browse to a file.
  * Extracts the args using datastack.py.
  */

  constructor(props) {
    super(props);
    this.browseFile = this.browseFile.bind(this);
  }

  async browseFile(event) {
    const data = await Electron.remote.dialog.showOpenDialog()
    const payload = { 
      datastack_path: data.filePaths[0]
    }
    const datastack = await fetchDatastackFromFile(payload)
    console.log(datastack)
    const specLoaded = await this.props.investGetSpec(datastack.module_name)
    console.log('after promised getspec, before batch update')
    console.log(specLoaded)
    if (specLoaded) { this.props.batchUpdateArgs(datastack['args']) }
  }

  render() {
    return(
      <Button 
        onClick={this.browseFile}
        variant="primary">
        Load Parameters
      </Button>
    );
  }
}

LoadButton.propTypes = {
  argsToJsonFile: PropTypes.func,
  disabled: PropTypes.bool
}

