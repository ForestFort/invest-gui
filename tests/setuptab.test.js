import fs from 'fs';
import path from 'path';
import React from 'react';
import { fireEvent, render,
         wait, waitForElement } from '@testing-library/react'
import '@testing-library/jest-dom'
import { SetupTab } from '../src/components/SetupTab';
import { fetchValidation } from '../src/server_requests';
jest.mock('../src/server_requests');
// generated this file from `invest getspec carbon --json`
// import ARGS_SPEC from './data/carbon_args_spec.json';

const ARG_TYPE_INPUT_MAP = {
  csv: "text",
  vector: "text",
  raster: "text",
  directory: "text",
  freestyle_string: "text",
  number: "text",
  boolean: "radio",
  option_string: "select"
}

function renderSetupFromSpec(spec) {
  const {getByText, getByLabelText, ...utils} = render(
    <SetupTab
      args={spec}
      argsValid={false}
      modulename={null}
      updateArg={() => {}}
      batchUpdateArgs={() => {}}
      investValidate={() => {}}
      argsValuesFromSpec={() => {}}
      investExecute={() => {}}
    />)
  fetchValidation.mockResolvedValue([[Object.keys(spec), 'invalid because']])
  return { utils, getByText, getByLabelText }
}

test('SetupTab: an input form for a directory', async () => {
  const spec = { arg: { name: 'Workspace', type: 'directory' } }
  const {utils, getByText, getByLabelText } = renderSetupFromSpec(spec)
  const input = getByLabelText(spec.arg.name)
  fireEvent.change(input, { target: { value: 'foo' } })
  utils.debug(input)
  await wait(() => {
    expect(input).toHaveValue('foo')
    expect(input).toHaveAttribute('type', 'text')
    // expect(input).toBeInvalid()
    expect(getByText('Browse'))
  })
})

test('SetupTab: an input form for a csv', async () => {
  const spec = { arg: { name: 'foo', type: 'csv' } }
  const {utils, getByText, getByLabelText } = renderSetupFromSpec(spec)
  await wait(() => {
    expect(getByLabelText(spec.arg.name)).toHaveAttribute('type', 'text')
    expect(getByText('Browse'))
  })
})

test('SetupTab: an input form for a vector', async () => {
  const spec = { arg: { name: 'foo', type: 'vector' } }
  const {utils, getByText, getByLabelText } = renderSetupFromSpec(spec)
  await wait(() => {
    expect(getByLabelText(spec.arg.name)).toHaveAttribute('type', 'text')
    expect(getByText('Browse'))
  })
})

test('SetupTab: an input form for a raster', async () => {
  const spec = { arg: { name: 'foo', type: 'raster' } }
  const {utils, getByText, getByLabelText } = renderSetupFromSpec(spec)
  await wait(() => {
    expect(getByLabelText(spec.arg.name)).toHaveAttribute('type', 'text')
    expect(getByText('Browse'))
  })
})

test('SetupTab: an input form for a freestyle_string', async () => {
  const spec = { arg: { name: 'foo', type: 'freestyle_string' } }
  const {utils, getByText, getByLabelText } = renderSetupFromSpec(spec)
  await wait(() => {
    expect(getByLabelText(spec.arg.name)).toHaveAttribute('type', 'text')
  })
})

test('SetupTab: an input form for a number', async () => {
  const spec = { arg: { name: 'foo', type: 'number' } }
  const {utils, getByText, getByLabelText } = renderSetupFromSpec(spec)
  await wait(() => {
    expect(getByLabelText(spec.arg.name)).toHaveAttribute('type', 'text')
  })
})

test('SetupTab: an input form for a boolean', async () => {
  const spec = { arg: { name: 'foo', type: 'boolean' } }
  const {utils, getByText, getByLabelText } = renderSetupFromSpec(spec)
  await wait(() => {
    expect(getByLabelText(spec.arg.name)).toHaveAttribute('type', 'radio')
  })
})

test('SetupTab: an input form for an option_string', async () => {
  const spec = { arg: { 
    name: 'foo', 
    type: 'option_string', 
    validation_options: { options: ['a', 'b'] } } }
  const {utils, getByText, getByLabelText } = renderSetupFromSpec(spec)
  await wait(() => {
    // utils.debug(getByLabelText(spec.arg.name))
    expect(getByLabelText(spec.arg.name)).toHaveValue('a');
    expect(getByLabelText(spec.arg.name)).not.toHaveValue('b');
  })
})
