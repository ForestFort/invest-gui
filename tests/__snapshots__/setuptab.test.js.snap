// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetupTab: a model input form 1`] = `
<div>
  <form
    className=""
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <div
      className="form-group row"
    >
      <label
        className="form-label col-form-label col-sm-3"
      >
        Workspace
      </label>
      <div
        className="col-sm-8"
      >
        <div
          className="input-group"
        >
          <input
            className="form-control"
            name="workspace_dir"
            onChange={[Function]}
            type="text"
            value=""
          />
          <div
            className="input-group-append"
          >
            <button
              className="btn btn-outline-secondary"
              disabled={false}
              name="workspace_dir"
              onClick={[Function]}
              type="button"
              value="directory"
            >
              Browse
            </button>
          </div>
          <div
            className="invalid-feedback"
          >
            directory : 
          </div>
        </div>
      </div>
    </div>
    <div
      className="form-group row"
    >
      <label
        className="form-label col-form-label col-sm-3"
      >
        File suffix
      </label>
      <div
        className="col-sm-4"
      >
        <input
          className="form-control"
          name="results_suffix"
          onChange={[Function]}
          type="text"
          value=""
        />
        <div
          className="invalid-feedback"
        >
          freestyle_string : 
        </div>
      </div>
    </div>
    <div
      className="form-group row"
    >
      <label
        className="form-label col-form-label col-sm-3"
      >
        Current Land Use/Land Cover
      </label>
      <div
        className="col-sm-8"
      >
        <div
          className="input-group"
        >
          <input
            className="form-control"
            name="lulc_cur_path"
            onChange={[Function]}
            type="text"
            value=""
          />
          <div
            className="input-group-append"
          >
            <button
              className="btn btn-outline-secondary"
              disabled={false}
              name="lulc_cur_path"
              onClick={[Function]}
              type="button"
              value="raster"
            >
              Browse
            </button>
          </div>
          <div
            className="invalid-feedback"
          >
            raster : 
          </div>
        </div>
      </div>
    </div>
    <div
      className="form-group row"
    >
      <label
        className="form-label col-form-label col-sm-3"
      >
        Calculate Sequestration
      </label>
      <div
        className="col-sm-8"
      >
        <div
          className="form-check form-check-inline"
        >
          <input
            checked={false}
            className="form-check-input"
            disabled={false}
            name="calc_sequestration"
            onChange={[Function]}
            type="radio"
            value="true"
          />
          <label
            className="form-check-label"
            title=""
            type="checkbox"
          >
            Yes
          </label>
        </div>
        <div
          className="form-check form-check-inline"
        >
          <input
            checked={true}
            className="form-check-input"
            disabled={false}
            name="calc_sequestration"
            onChange={[Function]}
            type="radio"
            value="false"
          />
          <label
            className="form-check-label"
            title=""
            type="checkbox"
          >
            No
          </label>
        </div>
      </div>
    </div>
    <div
      className="form-group row"
    >
      <label
        className="form-label col-form-label col-sm-3"
      >
        Future Landcover
      </label>
      <div
        className="col-sm-8"
      >
        <div
          className="input-group"
        >
          <input
            className="form-control"
            name="lulc_fut_path"
            onChange={[Function]}
            type="text"
            value=""
          />
          <div
            className="input-group-append"
          >
            <button
              className="btn btn-outline-secondary"
              disabled={false}
              name="lulc_fut_path"
              onClick={[Function]}
              type="button"
              value="raster"
            >
              Browse
            </button>
          </div>
          <div
            className="invalid-feedback"
          >
            raster : 
          </div>
        </div>
      </div>
    </div>
    <div
      className="form-group row"
    >
      <label
        className="form-label col-form-label col-sm-3"
      >
        REDD Scenario Analysis
      </label>
      <div
        className="col-sm-8"
      >
        <div
          className="form-check form-check-inline"
        >
          <input
            checked={false}
            className="form-check-input"
            disabled={false}
            name="do_redd"
            onChange={[Function]}
            type="radio"
            value="true"
          />
          <label
            className="form-check-label"
            title=""
            type="checkbox"
          >
            Yes
          </label>
        </div>
        <div
          className="form-check form-check-inline"
        >
          <input
            checked={true}
            className="form-check-input"
            disabled={false}
            name="do_redd"
            onChange={[Function]}
            type="radio"
            value="false"
          />
          <label
            className="form-check-label"
            title=""
            type="checkbox"
          >
            No
          </label>
        </div>
      </div>
    </div>
    <div
      className="form-group row"
    >
      <label
        className="form-label col-form-label col-sm-3"
      >
        REDD Policy)
      </label>
      <div
        className="col-sm-8"
      >
        <div
          className="input-group"
        >
          <input
            className="form-control"
            name="lulc_redd_path"
            onChange={[Function]}
            type="text"
            value=""
          />
          <div
            className="input-group-append"
          >
            <button
              className="btn btn-outline-secondary"
              disabled={false}
              name="lulc_redd_path"
              onClick={[Function]}
              type="button"
              value="raster"
            >
              Browse
            </button>
          </div>
          <div
            className="invalid-feedback"
          >
            raster : 
          </div>
        </div>
      </div>
    </div>
    <div
      className="form-group row"
    >
      <label
        className="form-label col-form-label col-sm-3"
      >
        Carbon Pools
      </label>
      <div
        className="col-sm-8"
      >
        <div
          className="input-group"
        >
          <input
            className="form-control"
            name="carbon_pools_path"
            onChange={[Function]}
            type="text"
            value=""
          />
          <div
            className="input-group-append"
          >
            <button
              className="btn btn-outline-secondary"
              disabled={false}
              name="carbon_pools_path"
              onClick={[Function]}
              type="button"
              value="csv"
            >
              Browse
            </button>
          </div>
          <div
            className="invalid-feedback"
          >
            csv : 
          </div>
        </div>
      </div>
    </div>
    <div
      className="form-group row"
    >
      <label
        className="form-label col-form-label col-sm-3"
      >
        Current Landcover Calendar Year
      </label>
      <div
        className="col-sm-4"
      >
        <input
          className="form-control"
          name="lulc_cur_year"
          onChange={[Function]}
          type="text"
          value=""
        />
        <div
          className="invalid-feedback"
        >
          number : 
        </div>
      </div>
    </div>
    <div
      className="form-group row"
    >
      <label
        className="form-label col-form-label col-sm-3"
      >
        Future Landcover Calendar Year
      </label>
      <div
        className="col-sm-4"
      >
        <input
          className="form-control"
          name="lulc_fut_year"
          onChange={[Function]}
          type="text"
          value=""
        />
        <div
          className="invalid-feedback"
        >
          number : 
        </div>
      </div>
    </div>
    <div
      className="form-group row"
    >
      <label
        className="form-label col-form-label col-sm-3"
      >
        Run Valuation Model
      </label>
      <div
        className="col-sm-8"
      >
        <div
          className="form-check form-check-inline"
        >
          <input
            checked={false}
            className="form-check-input"
            disabled={false}
            name="do_valuation"
            onChange={[Function]}
            type="radio"
            value="true"
          />
          <label
            className="form-check-label"
            title=""
            type="checkbox"
          >
            Yes
          </label>
        </div>
        <div
          className="form-check form-check-inline"
        >
          <input
            checked={true}
            className="form-check-input"
            disabled={false}
            name="do_valuation"
            onChange={[Function]}
            type="radio"
            value="false"
          />
          <label
            className="form-check-label"
            title=""
            type="checkbox"
          >
            No
          </label>
        </div>
      </div>
    </div>
    <div
      className="form-group row"
    >
      <label
        className="form-label col-form-label col-sm-3"
      >
        Price/Metric ton of carbon
      </label>
      <div
        className="col-sm-4"
      >
        <input
          className="form-control"
          name="price_per_metric_ton_of_c"
          onChange={[Function]}
          type="text"
          value=""
        />
        <div
          className="invalid-feedback"
        >
          number : 
        </div>
      </div>
    </div>
    <div
      className="form-group row"
    >
      <label
        className="form-label col-form-label col-sm-3"
      >
        Market Discount in Price of Carbon (%)
      </label>
      <div
        className="col-sm-4"
      >
        <input
          className="form-control"
          name="discount_rate"
          onChange={[Function]}
          type="text"
          value=""
        />
        <div
          className="invalid-feedback"
        >
          number : 
        </div>
      </div>
    </div>
    <div
      className="form-group row"
    >
      <label
        className="form-label col-form-label col-sm-3"
      >
        Annual Rate of Change in Price of Carbon (%)
      </label>
      <div
        className="col-sm-4"
      >
        <input
          className="form-control"
          name="rate_change"
          onChange={[Function]}
          type="text"
          value=""
        />
        <div
          className="invalid-feedback"
        >
          number : 
        </div>
      </div>
    </div>
  </form>
  <button
    className="btn btn-primary btn-lg"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    Execute
  </button>
</div>
`;
